homeassistant/components/mqtt/__init__.py

```diff
@@ -338,64 +338,55 @@ async def finish_dump(_: datetime) -> None:
-                raise ServiceValidationError(
-                    translation_domain=ex.translation_domain,
-                    translation_key=ex.translation_key,
-                    translation_placeholders=ex.translation_placeholders,
-                ) from ex
-
-            new_config: list[ConfigType] = config_yaml.get(DOMAIN, [])
-            platforms_used = platforms_from_config(new_config)
-            new_platforms = platforms_used - mqtt_data.platforms_loaded
-            await async_forward_entry_setup_and_setup_discovery(
-                hass, entry, new_platforms
+    async def _reload_config(call: ServiceCall) -> None:
+        """Reload the platforms."""
+        # Fetch updated manually configured items and validate
+        try:
+            config_yaml = await async_integration_yaml_config(
+                hass, DOMAIN, raise_on_failure=True
             )
-            # Check the schema before continuing reload
-            await async_check_config_schema(hass, config_yaml)
-
-            # Remove repair issues
-            _async_remove_mqtt_issues(hass, mqtt_data)
-
-            mqtt_data.config = new_config
-
-            # Reload the modern yaml platforms
-            mqtt_platforms = async_get_platforms(hass, DOMAIN)
-            tasks = [
-                entity.async_remove()
-                for mqtt_platform in mqtt_platforms
-                for entity in mqtt_platform.entities.values()
-                if getattr(entity, "_discovery_data", None) is None
-                and mqtt_platform.config_entry
-                and mqtt_platform.domain in RELOADABLE_PLATFORMS
-            ]
-            await asyncio.gather(*tasks)
-
-            for component in mqtt_data.reload_handlers.values():
-                component()
-
-            # Fire event
-            hass.bus.async_fire(f"event_{DOMAIN}_reloaded", context=call.context)
-
-        async_register_admin_service(hass, DOMAIN, SERVICE_RELOAD, _reload_config)
+        except ConfigValidationError as ex:
+            raise ServiceValidationError(
+                translation_domain=ex.translation_domain,
+                translation_key=ex.translation_key,
+                translation_placeholders=ex.translation_placeholders,
+            ) from ex
+
+        new_config: list[ConfigType] = config_yaml.get(DOMAIN, [])
+        platforms_used = platforms_from_config(new_config)
+        new_platforms = platforms_used - mqtt_data.platforms_loaded
+        await async_forward_entry_setup_and_setup_discovery(hass, entry, new_platforms)
+        # Check the schema before continuing reload
+        await async_check_config_schema(hass, config_yaml)
+
+        # Remove repair issues
+        _async_remove_mqtt_issues(hass, mqtt_data)
+
+        mqtt_data.config = new_config
+
+        # Reload the modern yaml platforms
+        mqtt_platforms = async_get_platforms(hass, DOMAIN)
+        tasks = [
+            entity.async_remove()
```

Comment:
This could be wrapped in `create_eager_task` if we do `list(mqtt_platform.entities.values())` to make sure it doesn't change during iteration